File: app/src/main/java/com/savvasdalkitsis/gameframe/control/presenter/ControlPresenter.java
package com.savvasdalkitsis.gameframe.control.presenter;

import com.savvasdalkitsis.gameframe.control.view.ControlView;
import com.savvasdalkitsis.gameframe.ip.repository.IpRepository;
import com.savvasdalkitsis.gameframe.model.Brightness;
import com.savvasdalkitsis.gameframe.model.ClockFace;
import com.savvasdalkitsis.gameframe.model.CycleInterval;
import com.savvasdalkitsis.gameframe.model.DisplayMode;
import com.savvasdalkitsis.gameframe.model.PlaybackMode;
import com.savvasdalkitsis.gameframe.rx.RxTransformers;
import com.savvasdalkitsis.gameframe.gameframe.usecase.GameFrameUseCase;

import rx.Observable;

public class ControlPresenter {

    private final IpRepository ipRepository;
    private ControlView controlView;
    private final GameFrameUseCase gameFrameUseCase;

    public ControlPresenter(GameFrameUseCase gameFrameUseCase, IpRepository ipRepository) {
        this.gameFrameUseCase = gameFrameUseCase;
        this.ipRepository = ipRepository;
    }

    public void bindView(ControlView controlView) {
        this.controlView = controlView;
    }

    public void loadIpAddress() {
        ipRepository.getIpAddress()
                .compose(RxTransformers.schedulers())
                .subscribe(controlView::ipAddressLoaded, controlView::ipCouldNotBeFound);
    }

    public void togglePower() {
        runCommandAndNotifyView(gameFrameUseCase.togglePower());
    }

    public void menu() {
        runCommandAndNotifyView(gameFrameUseCase.menu());
    }

    public void next() {
        runCommandAndNotifyView(gameFrameUseCase.next());
    }

    public void changeBrightness(Brightness brightness) {
        runCommandAndIgnoreResult(gameFrameUseCase.setBrightness(brightness));
    }

    public void changePlaybackMode(PlaybackMode playbackMode) {
        runCommandAndIgnoreResult(gameFrameUseCase.setPlaybackMode(playbackMode));
    }

    public void changeCycleInterval(CycleInterval cycleInterval) {
        runCommandAndIgnoreResult(gameFrameUseCase.setCycleInterval(cycleInterval));
    }

    public void changeDisplayMode(DisplayMode displayMode) {
        runCommandAndIgnoreResult(gameFrameUseCase.setDisplayMode(displayMode));
    }

    public void changeClockFace(ClockFace clockFace) {
        runCommandAndIgnoreResult(gameFrameUseCase.setClockFace(clockFace));
    }

    private Observable<Void> runCommand(Observable<Void> command) {
        return command.compose(RxTransformers.interceptIpMissingException())
                .compose(RxTransformers.schedulers());
    }

    private void runCommandAndNotifyView(Observable<Void> command) {
        runCommand(command).subscribe(n -> controlView.operationSuccess(), e -> controlView.operationFailure(e));
    }

    private void runCommandAndIgnoreResult(Observable<Void> command) {
        runCommand(command).subscribe(n -> {}, e -> {});
    }
}

File: app/src/main/java/com/savvasdalkitsis/gameframe/control/presenter/ControlPresenter.kt
package com.savvasdalkitsis.gameframe.control.presenter

import com.savvasdalkitsis.gameframe.control.model.*
import com.savvasdalkitsis.gameframe.control.view.ControlView
import com.savvasdalkitsis.gameframe.gameframe.usecase.GameFrameUseCase
import com.savvasdalkitsis.gameframe.ip.model.IpAddress
import com.savvasdalkitsis.gameframe.ip.repository.IpRepository
import com.savvasdalkitsis.gameframe.infra.rx.RxTransformers
import io.reactivex.Completable

class ControlPresenter(private val gameFrameUseCase: GameFrameUseCase, private val ipRepository: IpRepository) {

    private lateinit var controlView: ControlView

    fun bindView(controlView: ControlView) {
        this.controlView = controlView
    }

    fun loadIpAddress() {
        ipRepository.ipAddress
                .compose<IpAddress>(RxTransformers.schedulers<IpAddress>())
                .subscribe({ controlView.ipAddressLoaded(it) }, { controlView.ipCouldNotBeFound(it) })
    }

    fun togglePower() {
        runCommandAndNotifyView(gameFrameUseCase.togglePower())
    }

    fun menu() {
        runCommandAndNotifyView(gameFrameUseCase.menu())
    }

    fun next() {
        runCommandAndNotifyView(gameFrameUseCase.next())
    }

    fun changeBrightness(brightness: Brightness) {
        runCommandAndIgnoreResult(gameFrameUseCase.setBrightness(brightness))
    }

    fun changePlaybackMode(playbackMode: PlaybackMode) {
        runCommandAndIgnoreResult(gameFrameUseCase.setPlaybackMode(playbackMode))
    }

    fun changeCycleInterval(cycleInterval: CycleInterval) {
        runCommandAndIgnoreResult(gameFrameUseCase.setCycleInterval(cycleInterval))
    }

    fun changeDisplayMode(displayMode: DisplayMode) {
        runCommandAndIgnoreResult(gameFrameUseCase.setDisplayMode(displayMode))
    }

    fun changeClockFace(clockFace: ClockFace) {
        runCommandAndIgnoreResult(gameFrameUseCase.setClockFace(clockFace))
    }

    private fun runCommand(command: Completable) =
            command.compose(RxTransformers.interceptIpMissingException())
                    .compose(RxTransformers.schedulers())

    private fun runCommandAndNotifyView(command: Completable) {
        runCommand(command).subscribe({ controlView.operationSuccess() }, { e -> controlView.operationFailure(e) })
    }

    private fun runCommandAndIgnoreResult(command: Completable) {
        runCommand(command).subscribe({ }, { })
    }
}
